File: clients/ios/AtlasBot/WeComAuth.swift
import Foundation

/// WM1 WeComProvider — ASWebAuthenticationSession → code → Hub exchange → TokenStore.
///
/// Contract (aligned with PC `exchangeWeComCode` / W1):
/// - Authorize URL: corpId/agentId; **no PKCE** (state + one-time code).
/// - Callback remains `atlasbot://auth/callback`.
/// - Exchange: Hub `POST /auth/wecom/exchange` (secret Hub-only).
/// - Store Hub JWT in Keychain via `TokenStore`; `HubClient` Bearer on upgrade.
/// - **No WebView** primary path (ASWebAuthenticationSession only).
/// - Provider switch: `ATLAS_TICKET_PROVIDER=wecom`; default oidc.
///
/// See docs/i2-login-runbook.md § WeCom / § Mobile.
enum WeComAuth {
    static let provider = "wecom"
    static let redirectURI = "atlasbot://auth/callback"

    struct Config {
        var corpId: String
        var agentId: String
        var authorizeBase: String
        var hubHttpBase: String
        var redirectUri: String = WeComAuth.redirectURI
    }

    struct ExchangeResponse {
        var accessToken: String
        var tokenType: String?
        var expiresIn: Int64?
        var subject: String?
    }

    static func buildAuthorizeURL(cfg: Config, state: String) -> URL? {
        var root = cfg.authorizeBase
        while root.hasSuffix("/") { root.removeLast() }
        let authorize = root.hasSuffix("/authorize") ? root : root + "/authorize"
        var c = URLComponents(string: authorize)
        c?.queryItems = [
            URLQueryItem(name: "appid", value: cfg.corpId),
            URLQueryItem(name: "redirect_uri", value: cfg.redirectUri),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "scope", value: "snsapi_base"),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "agentid", value: cfg.agentId),
        ]
        c?.fragment = "wechat_redirect"
        return c?.url
    }

    static func subject(corpId: String, userid: String) -> String {
        "wecom:\(corpId):\(userid)"
    }

    static func exchangeURL(hubHttpBase: String) -> URL? {
        var root = hubHttpBase
        while root.hasSuffix("/") { root.removeLast() }
        return URL(string: root + "/auth/wecom/exchange")
    }

    /// POST Hub `/auth/wecom/exchange` — JSON `{code, state?}` → `access_token`.
    /// Client never holds `ATLAS_WECOM_SECRET`.
    static func exchangeCode(
        hubHttpBase: String,
        code: String,
        state: String? = nil,
        session: URLSession = .shared,
        completion: @escaping (Result<ExchangeResponse, Error>) -> Void
    ) {
        guard let url = exchangeURL(hubHttpBase: hubHttpBase) else {
            completion(.failure(NSError(domain: "wecom", code: 1, userInfo: [NSLocalizedDescriptionKey: "bad exchange url"])))
            return
        }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        var body: [String: String] = ["code": code]
        if let state, !state.isEmpty { body["state"] = state }
        do {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            completion(.failure(error))
            return
        }
        session.dataTask(with: req) { data, resp, err in
            if let err { completion(.failure(err)); return }
            let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
            let bodyData = data ?? Data()
            guard (200...299).contains(status) else {
                let text = String(data: bodyData, encoding: .utf8) ?? ""
                completion(.failure(NSError(domain: "wecom", code: status, userInfo: [NSLocalizedDescriptionKey: "wecom exchange \(status): \(text)"])))
                return
            }
            do {
                completion(.success(try parseExchangeJSON(bodyData)))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }

    static func parseExchangeJSON(_ data: Data) throws -> ExchangeResponse {
        let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        guard let token = obj["access_token"] as? String, !token.isEmpty else {
            throw NSError(domain: "wecom", code: 2, userInfo: [NSLocalizedDescriptionKey: "exchange missing access_token"])
        }
        return ExchangeResponse(
            accessToken: token,
            tokenType: obj["token_type"] as? String,
            expiresIn: (obj["expires_in"] as? NSNumber)?.int64Value,
            subject: obj["subject"] as? String
        )
    }
}

/// Derive Hub HTTP base from a Hub WS URL.
public func wsToHttpBase(_ ws: String) -> String {
    var s = ws.trimmingCharacters(in: .whitespacesAndNewlines)
    if s.hasPrefix("ws://") { s = "http://" + s.dropFirst(5) }
    else if s.hasPrefix("wss://") { s = "https://" + s.dropFirst(6) }
    if s.hasSuffix("/ws") { s = String(s.dropLast(3)) }
    while s.hasSuffix("/") { s.removeLast() }
    return s
}

/// `ATLAS_TICKET_PROVIDER` → oidc|wecom (default oidc).
public func ticketProviderFromEnv(_ raw: String?) -> String {
    let v = (raw ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    if v == WeComAuth.provider { return WeComAuth.provider }
    return "oidc"
}

File: clients/android/app/src/main/java/com/atlasbot/client/auth/WeComAuth.kt
package com.atlasbot.client.auth

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

/**
 * WM1 WeComProvider — system browser (Custom Tabs) → code → Hub exchange → TokenStore.
 *
 * Contract (aligned with PC `exchangeWeComCode` / W1):
 * - Authorize URL uses corpId/agentId (**no PKCE**; state + one-time code).
 * - Callback remains [MOBILE_REDIRECT_URI] (`atlasbot://auth/callback`).
 * - Exchange is **Hub** `POST /auth/wecom/exchange` (secret never in the APK).
 * - Store Hub JWT via [TokenStore]; [HubClient] sends `Authorization: Bearer`.
 * - **No WebView** primary path (Custom Tabs / system browser only).
 * - Switch with `ATLAS_TICKET_PROVIDER=wecom` (or UI advanced); default stays oidc.
 *
 * See docs/i2-login-runbook.md § WeCom / § Mobile.
 */
object WeComAuth {
    const val PROVIDER = "wecom"

    data class Config(
        val corpId: String,
        val agentId: String,
        /** mock-wecom base or live open.weixin authorize host path */
        val authorizeBase: String,
        val hubHttpBase: String,
        val redirectUri: String = MOBILE_REDIRECT_URI,
    )

    data class ExchangeResponse(
        val accessToken: String,
        val tokenType: String? = null,
        val expiresIn: Long? = null,
        val subject: String? = null,
    )

    fun buildAuthorizeUrl(cfg: Config, state: String): String {
        val base = cfg.authorizeBase.trimEnd('/')
        val authorize = if (base.endsWith("/authorize")) base else "$base/authorize"
        val q = listOf(
            "appid" to cfg.corpId,
            "redirect_uri" to cfg.redirectUri,
            "response_type" to "code",
            "scope" to "snsapi_base",
            "state" to state,
            "agentid" to cfg.agentId,
        ).joinToString("&") { (k, v) ->
            "$k=${java.net.URLEncoder.encode(v, Charsets.UTF_8.name())}"
        }
        return "$authorize?$q#wechat_redirect"
    }

    fun subject(corpId: String, userid: String): String = "wecom:$corpId:$userid"

    fun exchangeUrl(hubHttpBase: String): String =
        "${hubHttpBase.trimEnd('/')}/auth/wecom/exchange"

    /**
     * POST Hub `/auth/wecom/exchange` — JSON `{code, state?}` → `access_token` (Hub JWT).
     * Client never holds `ATLAS_WECOM_SECRET`. Blocking — call off main thread.
     */
    fun exchangeCode(
        hubHttpBase: String,
        code: String,
        state: String? = null,
    ): ExchangeResponse {
        val url = exchangeUrl(hubHttpBase)
        val bodyJson = JSONObject().apply {
            put("code", code)
            if (!state.isNullOrEmpty()) put("state", state)
        }
        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            doOutput = true
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Content-Type", "application/json")
            connectTimeout = 30_000
            readTimeout = 30_000
        }
        conn.outputStream.use { it.write(bodyJson.toString().toByteArray(StandardCharsets.UTF_8)) }
        val status = conn.responseCode
        val body = (if (status in 200..299) conn.inputStream else conn.errorStream)
            ?.bufferedReader()?.readText().orEmpty()
        if (status !in 200..299) {
            throw IllegalStateException("wecom exchange $status: $body")
        }
        return parseExchangeJson(body)
    }

    fun parseExchangeJson(body: String): ExchangeResponse {
        val o = JSONObject(body)
        val token = o.optString("access_token").takeIf { it.isNotEmpty() }
            ?: throw IllegalStateException("exchange missing access_token")
        return ExchangeResponse(
            accessToken = token,
            tokenType = o.optString("token_type").takeIf { it.isNotEmpty() },
            expiresIn = if (o.has("expires_in") && !o.isNull("expires_in")) o.optLong("expires_in") else null,
            subject = o.optString("subject").takeIf { it.isNotEmpty() },
        )
    }
}

/** Derive Hub HTTP base from a Hub WS URL (`ws(s)://host/ws` → `http(s)://host`). */
fun wsToHttpBase(ws: String): String {
    var s = ws.trim()
    when {
        s.startsWith("ws://") -> s = "http://" + s.removePrefix("ws://")
        s.startsWith("wss://") -> s = "https://" + s.removePrefix("wss://")
    }
    if (s.endsWith("/ws")) s = s.dropLast(3)
    return s.trimEnd('/')
}

/** `ATLAS_TICKET_PROVIDER` → oidc|wecom (default oidc). */
fun ticketProviderFromEnv(raw: String?): String {
    val v = raw?.trim()?.lowercase().orEmpty()
    return if (v == WeComAuth.PROVIDER) WeComAuth.PROVIDER else "oidc"
}
